### PR TITLE
Fix perf regression from #113569

### PR DESCRIPTION
**Pull request by RalfJung**
Fix perf regression from #113569

r? @oli-obk

### CODE REVIEW

**Comment by RalfJung** [comment]
That seems to be most of the regressions, nice.

Now let's see which part it is...

**Comment by RalfJung** [comment]
Looks like the write_uninit does cause a noticeable regression. It could also help UB detection, but this kind of UB is somewhat unlikely -- I think you can't even trigger it without custom MIR.

@bors try @rust-timer queue
@oli-obk this is ready for review.

**Comment by RalfJung** [comment]
Uh, what? Why did the perf changes disappear?

File: compiler/rustc_const_eval/src/interpret/machine.rs [hunk 1]
```diff
@@ -421,13 +421,17 @@ pub trait Machine<'mir, 'tcx: 'mir>: Sized {
...
-    /// argument/return value was actually copied or passed in-place..
+    /// argument/return value was actually copied or passed in-place.
     fn protect_in_place_function_argument(
         ecx: &mut InterpCx<'mir, 'tcx, Self>,
         place: &PlaceTy<'tcx, Self::Provenance>,
     ) -> InterpResult<'tcx> {
         // Without an aliasing model, all we can do is put `Uninit` into the place.
-        ecx.write_uninit(place)
+        // This can only be violated with custom MIR though so we avoid the perf hit.
+        if ecx.tcx.sess.opts.unstable_opts.extra_const_ub_checks {
```
Comment (pnkfelix) [inline compiler/rustc_const_eval/src/interpret/machine.rs:431]: @RalfJung , this flag has to be turned on explicitly, right?

Can/should we also unconditionally apply this logic when the code being compiled was defined via `#[custom_mir(..)]` ?

Comment (RalfJung) [reply compiler/rustc_const_eval/src/interpret/machine.rs:431]: Indeed this is an off-by-default flag, also guarding whether we check for value validity on every typed copy.

I feel like even with custom MIR, causing UB here is so unlikely that it's not worth the trouble. However, for some reason it seems that just checking this flag is already sufficient to mostly negate all perf benefits? Or the perf changes we saw earlier were mostly noise.

**Comment by RalfJung** [comment]
I don't see a clear error message, no idea what is going on
```
 Error: Optimized build pipeline has failed

Caused by:
    Command RUSTC=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/rustc RUSTC_BOOTSTRAP=1 /checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo build -p collector [at /tmp/tmp-multistage/opt-artifacts/rustc-perf] has failed with exit code Some(101)
```

**Comment by RalfJung** [comment]
Even the full version of this PR (identical to what I tested earlier) cannot find any perf benefits any more. The perf triage in https://github.com/rust-lang/rust/pull/113569 concluded this is likely mostly spurious. Hence closing this PR.

**Pull request closed without merging**